lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -2758,15 +2821,22 @@ ConformanceChecker::resolveWitnessViaLookup(ValueDecl *requirement) {
           auto &ctx = witness->getASTContext();
           auto &diags = ctx.Diags;
           {
+            SourceLoc diagLoc = getLocForDiagnosingWitness(conformance,witness);
             auto diag = diags.diagnose(
-                          witness,
-                          hasAnyError(adjustments)
-                            ? diag::err_protocol_witness_optionality
-                            : diag::warn_protocol_witness_optionality,
-                          classifyOptionalityIssues(adjustments, requirement),
-                          witness->getFullName(),
-                          proto->getFullName());
-            addOptionalityFixIts(adjustments, ctx, witness, diag);
+                diagLoc,
+                hasAnyError(adjustments)
+                  ? diag::err_protocol_witness_optionality
+                  : diag::warn_protocol_witness_optionality,
+                classifyOptionalityIssues(adjustments, requirement),
+                witness->getFullName(),
+                proto->getFullName());
+            if (diagLoc == witness->getLoc()) {
+              addOptionalityFixIts(adjustments, ctx, witness, diag);
```

Comment:
What's the thinking behind these only applying in this case, rather than the `reset`/`emplace` thing you've done for other similar versions to get fixits on the note?

Reply:
I *mostly* only did that where there was a dedicated note, rather than just a "declared here". There are two reasons for that: "declared here" doesn't tell you what the fix-it is doing, and if the declaration is written somewhere else, it's less likely that its sole purpose is to satisfy this protocol. The one exception was the `required` initializer fix-it, because people complain that they never remember what they have to write for initializers but I couldn't think of a good custom note for that case.

Happy to accept feedback here.

Reply:
Sounds like reasonable logic.